eclass/dist-kernel-utils.eclass

```diff
@@ -169,20 +198,15 @@ dist-kernel_compressed_module_cleanup() {
 
 	[[ ${#} -ne 1 ]] && die "${FUNCNAME}: invalid arguments"
 	local path=${1}
-	local basename f
+	local compress=$(dist-kernel_detect_compressor "${path}/source")
+	local basename suffix
 
 	while read -r basename; do
-		local prev=
-		for f in "${path}/${basename}"{,.gz,.xz,.zst}; do
-			if [[ ! -e ${f} ]]; then
+		for suffix in .ko .ko.gz .ko.xz .ko.zst; do
```

Comment:
I don't think you account for the possibility that there is no *new* compressed file, e.g. when you're reinstalling the same kernel package but modules were installed by other packages. You must only do cleanup when at least two matching files exist.

Reply:
That part is taken care off by `| sort | uniq -d` below, we only run this `while` loop for files that, with the suffix removed, have the same filename. So if there is no replacing module, the old module is kept.

Reply:
Hmm, perhaps you're right. Still, I'd feel better having some tests.

Reply:
Ideally, test various `.config`s with a tree containing modules like:

```
a.ko
b.ko.gz
c.ko.xz
d.ko
d.ko.gz
e.ko
e.ko.xz
f.ko.gz
f.ko.xz
```

Something like that.

Reply:
Hmm, actually after thinking about it some more. This does indeed not account for the case where e.g. `foo.ko` and `foo.ko.gz` exist, while we are installing modules with `.ko.xz` but `foo.ko.xz` is not installed. Both `foo.ko` and `foo.ko.gz` would be removed. 

I propose that, if the module with the compression we want exists then we remove all others, if not then the duplicate module was not part of the install tree and we fallback to the old logic of keeping whichever module variant is the newest.

Reply:
Added tests for `dist-kernel_compressed_module_cleanup`.

Reply:
Hmm, still not 100% sure this is the most correct behavior. Think of the following degenerate case: you install `=foo-1` with `modules-compress`, then `=foo-2` without. For some reason, cleanup didn't happen and you end up with *new* `foo.ko` and *old* `foo.ko.xz`. FWIU the new logic would keep the old file.

Reply:
That won't happen because if `modules-compress` is not in `IUSE` or if the flag is disable, then `dist-kernel_detect_compressor` returns `.ko` (regardless of the `MODULE_COMPRESS_*` setting in the kernel config).

So in your example case, since the flag for `=foo-2` is disabled, `.ko` is kept and `.ko.xz` is removed.

Reply:
I'm talking about modules that were installed by *another* package than the one that's being run right now.

Reply:
I see, yes then it would remove the modules based on the settings of the package currently being merged.

I don't really see a way to resolve that problem though. If the cleanup detects duplicates that do not belong to our package then I think it should still remove the duplicates, leaving them in the module tree will probably confuse e.g. `depmod`/`kmod`. Therefore we must somehow decide which variation to remove, `-nt` is not a good criteria because of binpkgs, and the settings for the current package do not necessarily reflect those of the package owning the module. I honestly don't think there is any criteria we can use here that will guarantee the proper variant is removed since we cannot know about the settings of the owning package without querying who owns a module for every single module (which would be very slow).

However, given that this will only occur if:
- something has gone wrong on a previous cleanup, and
- despite the error message shown during the previous cleanup, the user ignored this and did not re-emerge the original package where the failure occurred, and
- USE=+/-modules-compress is not globally applied,
I think this situation will be extremely rare. 

The rarity, combined with that we cannot really reliably resolve this, makes me think that we can honestly just ignore this. If there was some failure in `foo`, then the user should really just re-emerge `foo` and then this will not occur.

Besides, this problem where the cleanup touches modules that do not necessarily belong to the package being merged already exists in the current code. So it really is tangential to the problem at hand (module cleanup from binpkgs being broken).

Reply:
Yeah, I guess we can ignore this. I wonder if it could be simpler to: 1) keep the correct compression if it exists, 2) keep `.ko` if it exists, 3) keep random one if neither exists.

Reply:
> Yeah, I guess we can ignore this. I wonder if it could be simpler to: 1) keep the correct compression if it exists, 2) keep `.ko` if it exists, 3) keep random one if neither exists.

In terms of code that would be simpler. However, I think that falling back on the `-nt` based check when the module with the desired compression does not exist is more "correct". Since this will usually produce the desired result, it will only be incorrect if binpkgs are used with old build times and the list of installed modules is changed at the same time. 

Whereas hardcoding that we prefer uncompressed over compressed when the desired compression does not exist is a bit more arbitrary.

Reply:
My logic is that the kernel is always able to load plain `.ko` whereas it is not able to load modules compressed with a different algorithm AFAIR.

Reply:
That said, I'm fine either way.